src/espidf.rs

```diff
@@ -105,47 +94,87 @@ impl Tools {
     }
 }
 
-/// Installer for the esp-idf source and tools.
+/// Information about a esp-idf source and tools installation.
+pub struct EspIdf {
+    /// The esp-idf repository.
+    pub esp_idf: git::Repository,
+    /// The binary paths of all tools concatenated with the system `PATH` env variable.
+    pub exported_path: OsString,
+}
+
+impl EspIdf {
+    pub fn detect_from_path() -> Result<Option<Self>> {
+        if cmd!["python", "idf.py", "--help"].is_ok() {
+            let version =
+                EspIdfVersion::try_from_env_var()?.ok_or_else(|| anyhow::anyhow!("TODO"))?;
+
+            let idf = Self {
+                esp_idf: version,
+                exported_path: env::var_os("PATH").unwrap_or_else(OsString::new),
+            };
+
+            Ok(Some(idf))
+        } else {
+            Ok(None)
+        }
+    }
```

Comment:
> I think a better way to detect this is:
> 
> * Move the logic for the `IDF_PATH` env var here, find out if it is a valid git repo.

But then also see my comments why I think doing overcomplicated checks for an activated ESP-IDF environment is not so ideal either. BTW "moving" the `IDF_PATH` logic to here would not do, as the same logic is necessary (and called from) `esp-idf-sys` - see below.

>   * See my comment below, why I think this is better.
>   * If this is not the case we're definitely not in an activated esp-idf environment.
> * Use the `which` crate to get the path to the first `python` in the `PATH` var.
>   
>   * I guess the edge-case for this is differences between the OSes (especially windows)
>     
>     * Windows can "hook" into calling python and use its own version (or just open Store) even if python is in `PATH`.
>       Though I think this is somewhat rare, but why not do this when doing this for `idf.py` already.
>   * Maybe make sure this python satisfies `${IDF_PATH}/requirements.txt`?
> * Use the `which` crate to see if there is a `idf.py` in the `PATH` and to get its path.
>   
>   * This prevents the edge-case where a file `idf.py` is in the current working directory.
>   * Make sure `idf.py` is in `${IDF_PATH}/tools`.
> * (Finally, we can try to run `idf.py --version` (with the retrieved `python`) to find out if it actually works.) Though this shouldn't be needed anymore, as we already verified `python` and the esp-idf git repo.

Some of these I might incorporate. Let me dig a bit further into these details, and the `which` crate. But then again and in general - I believe we should not do something extremely complex here, as mis-configured ESP-IDF environment, including an `idf.py` in the local directly would be very rare cases.

> I think the return value should just be a `Result`: either we succeeded in detecting an installation from `PATH` or failed in doing so. If we want to delineate why it failed we could for example define our own error enum with the `thiserror` crate:
> 
> * `NotFound`: When no esp-idf repo was found.
> * `Invalid`: When there is a esp-idf repo but either `python` or `idf.py` is invalid.
> * `Other`: Other system errors that may happen.
> 
> With this, the user of this function can decide how strict it should be in regards to failing or defaulting to installing the tools automatically.

With the rename of `detect_from_path` to `try_from_env` that I suggested elsewhere, I actually prefer the current signature:
* `try_` means the function is OK to return nothing, hence the `Option`
* The function can fail for other reasons, hence the `Result<Option<X>>` in the end

If I keep `detect_from_path` as a name then I agree, the current return type is awkward. Also in general, I don't think returning `Result<Option>` is an anti-pattern in Rust, and I cannot see indications of it?

Reply:
> * `try_` means the function is OK to return nothing, hence the `Option`

I think `try_` usually just means the function will return a `Result`, i.e. it can return an error (comparable to `try catch` in other languages).


> Also in general, I don't think returning `Result<Option>` is an anti-pattern in Rust, and I cannot see indications of it?

I just think having to deal if a `Result` is enough. And I guess I'm arguing when no esp-idf is detected from the env it should instead be `we failed to detect from env` (i.e. an `Err` value) and not `we succeeded in detecting nothing` (`Ok(None)`).